### PR TITLE
My coding challenge submission

### PR DESCRIPTION
**Pull request by gibsorya**
Here is my final solution for the coding challenge. Thanks for the template!

Would it be better if this gets put into a separate branch instead of main?

### CODE REVIEW

**Comment by gburgett** [comment]
Thanks Ryan!  I'll take a look at it shortly.

We intend to evaluate it then delete it without merging it, so we can keep the template clean for the next applicant!

File: app/controllers/messages_controller.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        else
+            render 'new'
+        end
+    end
+
+    def index
+        params.each do |key,value|
+            Rails.logger.warn "Param #{key}: #{value}"
+            if key == "tag"
+                @messages = Message.joins(:tags).where('tags.name' => value).group(:id)
```
Comment (gburgett) [inline app/controllers/messages_controller.rb:35]: Good solution here!  Extra credit question - can you tell me the (approximate, pseudocode) SQL query that this generates?  Just reply to my comment.

If not thats OK, just trying to get a gague on where you're at with SQL knowledge

**Review by gburgett — APPROVED** [review]
For the most part this is really good!  There's a couple of things you missed, which is OK for a coding challenge submission like this.  Those things would be caught in code review generally.  This was really helpful for us to get a gauge on how well you follow requirements and how well you can learn unfamiliar technologies.

I'll get with Justin and we'll email you with times to schedule the next interview!

Just for your own feedback, here's some minor things you missed:

1. The "content" field should be called "description" in the Form label.  The field name used in the DB doesn't matter nearly so much, what matters is presentation to the end user.
2. The pages had a consistent style but didn't use a grid system from a known CSS framework, the styles are all ad-hoc.  Not the "wrong" answer, but there's value in following a consistent well-known framework (this was in one of the requirements).  We use Bootstrap across all our apps at Watermark.
3. Copy-pasting the URL of the video into the URL bar causes the video to download instead of just playing it in the browser.  This has to do with setting attachment headers properly, which you left up to the ActiveStorage framework.  Leaving it up to the framework is the right answer 99% of the time, but in this case there was a specified requirement that the video plays instead of downloading (I could probably make that requirement more clear).

and one major thing was missed:

hitting the http://localhost:3000/messages without any tags causes a 500 error, both in the HTML view and the JSON.  This would have been caught if you did any controller specs.

Again, all of the above is just for your education and we appreciate the work you put into this!  I hope it was a good learning experience for you and we're looking forward to the next phase of the interview process!

File: app/controllers/messages_controller.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        #     # format.json  {render :json => @message}
+        # end
+    end
+    
+    def create
+        @message = Message.new(message_params)
+        @message.thumbnail.attach(params[:message][:thumbnail])
+        @message.video.attach(params[:message][:video])
+        if @message.save
+            reset_session
```
Comment (gburgett) [inline app/controllers/messages_controller.rb:23]: Would you please walk me through why you call reset_session here?  What is this intended to do?

Comment (gibsorya) [reply app/controllers/messages_controller.rb:23]: Honestly that's just a mistake that slipped by

**Comment by gibsorya** [comment]
Thanks for the feedback!

1. That's my bad.
2. I'm not too familiar with bootstrap but I know generally how it works. I'm willing to learn it though!
3. Interesting how it doesn't work for you - I can display the videos in my browser with the link no problem. I did have that issue when I first got the videos up and running though. Not sure why that isn't working for you
firefox_GElqh79MFK

With that `/messages` error, I just now am realizing I had that working but made changes to the `def index` method in the messages controller so I could get `?tag=` to work, but I forgot to make /messages working again. In that case, the way to fix that goes like this:
```Ruby    
def index
        params.each do |key,value|
            Rails.logger.warn "Param #{key}: #{value}"
            if key == "tag"
                @messages = Message.joins(:tags).where('tags.name' => value).group(:id)
                return
            end
        end
        @messages = Message.all
end
```
Having a spec for the controller would have definitely caught this issue.

File: app/controllers/messages_controller.rb (hunk 1 above)
Comment (gibsorya) [reply app/controllers/messages_controller.rb:35]: Good question! I believe that `SELECT "messages" * FROM "messages" INNER JOIN "taggings" WHERE "tag.name" = (insert tag name)` is close to the SQL query. I think it also has to find the message.ID and tag.ID in the "taggings" table, and then does another INNER JOIN with the "tags" table. I don't remember exactly how that works. I'd probably have to do a refresher course on SQL - it's been a while.

Comment (gburgett) [reply app/controllers/messages_controller.rb:35]: Great, and where does the `group(:id)` come in?

Comment (gibsorya) [reply app/controllers/messages_controller.rb:35]: Does it group the messages by message id? `GROUP BY`

**Comment by gibsorya** [comment]
¯\_(ツ)_/¯

**Pull request closed without merging**